### PR TITLE
feat: implement new node matcher API

### PR DESCRIPTION
**Pull request by gvozdvmozgu**

### CODE REVIEW

**Comment by gvozdvmozgu** [comment]
For this, we need a separate counter. We also have a problem where the `cache_key` is created for each instance, but currently, we don't store the instances; they are created and immediately dropped. This is about grammar, and it is resolved in [this pull request](https://github.com/quarylabs/sqruff/pull/370).

**Comment by gvozdvmozgu** [comment]
Oh, I wanted to comment on a different PR

File: crates/lib/src/core/parser/segments/base.rs [hunk 1]
```diff
@@ -734,7 +724,12 @@ impl Eq for ErasedSegment {}
 
 impl Hash for dyn Segment {
     fn hash<H: Hasher>(&self, state: &mut H) {
-        self.type_name().hash(state);
+        // TODO: remove hack
```
Comment (benfdking) [inline crates/lib/src/core/parser/segments/base.rs]: I think this can be removed?

**Review by benfdking — CHANGES_REQUESTED** [review]
Amazing work! I think there's one TODO that we can clean up? If not let me know.

**Review by benfdking — APPROVED** [review]

**Pull request merged**